### PR TITLE
Feature/pipline cli/rpc

### PR DESCRIPTION
**Pull request by yarinshaked**

### CODE REVIEW

File: pipert2/core/base/pipe.py [hunk 1]
```diff
@@ -45,6 +47,10 @@ def __init__(self, network: Network = QueueNetwork(),
             routine_synchroniser (RoutinesSynchroniser): Routine synchroniser if auto_pacing_mechanism is true, else None.
 
         """
+        super().__init__()
+
```
Comment (MayoG) [inline pipert2/core/base/pipe.py:51]: Remove please

File: pipert2/core/base/pipe.py [hunk 2]
```diff
@@ -45,6 +47,10 @@ def __init__(self, network: Network = QueueNetwork(),
...
+        self.rpc_listener = RPCListener(self)
+        self.rpc_server = zerorpc.Server(self.rpc_listener)
```
Comment (MayoG) [inline pipert2/core/base/pipe.py:53]: What if we don't want the server ?

Comment (MayoG) [reply pipert2/core/base/pipe.py:53]: Maybe add more options for choosing the endpoint and a way to start the server.

**Review by MayoG — CHANGES_REQUESTED** [review]

File: demo/rpc_client_demo.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import zerorpc
+from const import RPC_ENDPOINT
+
+if __name__ == '__main__':
+    client = zerorpc.Client()
+    client.connect(RPC_ENDPOINT)
+    print(client.pause())
```
Comment (MayoG) [inline demo/rpc_client_demo.py]: Please add all of the rpc scripts to a different single directory in the demo.

File: demo/rpc_server_pipe.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import time
```
Comment (MayoG) [inline demo/rpc_server_pipe.py:1]: Please add all of the rpc scripts to a single different directory in demos.

File: pipert2/core/base/pipe.py [hunk 5]
```diff
@@ -35,17 +35,7 @@ def __init__(self, network: Network = QueueNetwork(),
...
-        Attributes:
-            network: Network object responsible for the routine's communication.
-            logger: Logger object for logging the pipe actions.
-            data_transmitter: DataTransmitter object to indicate how data flows through the pipe by default.
-            flows (dict[str, Flow]): Dictionary mapping the pipe flows to their name.
-            event_board (EventBoard): EventBoard object responsible for the pipe events.
-            routine_synchroniser (RoutinesSynchroniser): Routine synchroniser if auto_pacing_mechanism is true, else None.
-
         """
-
```
Comment (MayoG) [inline pipert2/core/base/pipe.py:48]: Why did you delete our docs ??? :thinking:

File: pipert2/core/base/rpc_pipe_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+from zerorpc import Server
```
Comment (MayoG) [inline pipert2/core/base/rpc_pipe_wrapper.py:1]: Please create a new directory for pipe wrappers for future wrappers.

File: pipert2/core/base/rpc_pipe_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+from pipert2.core.base.pipe import Pipe
+from pipert2.utils.consts import START_EVENT_NAME, STOP_EVENT_NAME
+
+
+class RPCPipeWrapper:
```
Comment (MayoG) [inline pipert2/core/base/rpc_pipe_wrapper.py:6]: You can make the class inherit from the zerorpc server instead of holding it. [see here](https://github.com/0rpc/zerorpc-python/blob/master/tests/test_server.py#L42)
I think you should change it to that.

(Don't forget to update the demos if you decide to do that)

File: tests/unit/pipert/core/base/test_pipe.py [hunk 8]
```diff
@@ -1,22 +1,28 @@
...
     with patch('pipert2.core.base.pipe.Flow'):
         with patch('pipert2.core.base.pipe.EventBoard'):
             pipe = Pipe(network=Mock(), logger=Mock())
             yield pipe
 
 
-@pytest.fixture()
+@pytest.fixture
+def pipe_with_cli():
+    return Pipe(network=Mock(), logger=Mock(), run_cli=True, rpc_args={'endpoint': 'tcp://0.0.0.0:4242'})
```
Comment (MayoG) [inline tests/unit/pipert/core/base/test_pipe.py:22]: Please remove tests that are not used anymore.

**Review by MayoG — CHANGES_REQUESTED** [review]

File: pipert2/core/base/wrappers/rpc_pipe_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+from zerorpc import Server
+from pipert2.core.base.pipe import Pipe
+from pipert2.utils.consts import START_EVENT_NAME, STOP_EVENT_NAME
+
+
+class RPCPipeWrapper(Server):
+    """A RPC wrapper which invokes pipeline cli commands
```
Comment (eilonshi) [inline pipert2/core/base/wrappers/rpc_pipe_wrapper.py:7]: Change the way you implement this -> Instead of wrapping the pipe, the pipe will hold it

**Review by eilonshi — CHANGES_REQUESTED** [review]

File: pipert2/core/base/pipe.py [hunk 10]
```diff
@@ -17,49 +20,59 @@
...
             auto_pacing_mechanism: True if the user want to use auto pacing mechanism.
-
-        Attributes:
-            network: Network object responsible for the routine's communication.
-            logger: Logger object for logging the pipe actions.
-            data_transmitter: DataTransmitter object to indicate how data flows through the pipe by default.
-            flows (dict[str, Flow]): Dictionary mapping the pipe flows to their name.
-            event_board (EventBoard): EventBoard object responsible for the pipe events.
-            routine_synchroniser (RoutinesSynchroniser): Routine synchroniser if auto_pacing_mechanism is true, else None.
-
```
Comment (tomerdayan168) [inline pipert2/core/base/pipe.py:46]: Bravo :+1:

File: pipert2/core/base/pipe.py [hunk 11]
```diff
@@ -17,49 +20,59 @@
...
         else:
             self.routine_synchroniser = None
 
+    def run_rpc_server(self, endpoint: str):
+        """Binds it to a given endpoint and runs the rpc server.
+
+            Arguments:
+                endpoint: server's endpoint
+        """
+        self.rpc_server.run_rpc_server(endpoint=endpoint)
```
Comment (tomerdayan168) [inline pipert2/core/base/pipe.py:74]: I know it's strange, but if a user didn't pass True to use the RPC and call this function, then raise a custom error message or don't call the self.rpc_sever (because it is None instance) in order to avoid pipeline exit (choose what to do yourself, both are fine).

File: pipert2/core/base/wrappers/rpc_pipe_wrapper.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        """Invokes the start event in the pipe
+
+        """
+        self.notify_callback(event_name=START_EVENT_NAME)
+
+    def pause(self):
+        """Invokes the stop event in the pipe
+
+        """
+        self.notify_callback(event_name=STOP_EVENT_NAME)
```
Comment (tomerdayan168) [inline pipert2/core/base/wrappers/rpc_pipe_wrapper.py:39]: Delete

File: pipert2/core/base/wrappers/rpc_pipe_wrapper.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+    def kill(self):
+        """Invokes the kill event in the pipe
+
+        """
+        self.notify_callback(event_name="join", to_kill=True)
+        self.stop()
```
Comment (tomerdayan168) [inline pipert2/core/base/wrappers/rpc_pipe_wrapper.py:46]: Move the kill to the execute with the same logic. Maybe get another variable of "to_kill"

File: tests/unit/pipert/core/base/test_pipe.py [hunk 14]
```diff
@@ -1,22 +1,23 @@
 import pytest
+import zerorpc
+
```
Comment (tomerdayan168) [inline tests/unit/pipert/core/base/test_pipe.py:3]: Remove /n

File: pipert2/core/base/pipe.py [hunk 15]
```diff
@@ -4,8 +4,11 @@
...
 from pipert2.core.base.wire import Wire
 from pipert2.core.base.routine import Routine
+from pipert2.core.base.wrappers.rpc_pipe_wrapper import RPCPipeWrapper
 from pipert2.core.managers.network import Network
 from pipert2.core.managers.event_board import EventBoard
+from pipert2.utils.annotations import class_functions_dictionary
+from pipert2.utils.base_event_executor import BaseEventExecutor
 from pipert2.utils.consts.event_names import KILL_EVENT_NAME
 from pipert2.core.base.data_transmitter import DataTransmitter
 from pipert2.core.managers.networks.queue_network import QueueNetwork
```
Comment (tomerdayan168) [inline pipert2/core/base/pipe.py:14]: Keep the import sorted by length!! Elon will be proud

**Review by tomerdayan168 — CHANGES_REQUESTED** [review]
nice

File: pipert2/core/base/wrappers/rpc_pipe_wrapper.py (hunk 12 above)
Comment (yarinshaked) [reply pipert2/core/base/wrappers/rpc_pipe_wrapper.py:39]: basof lo

File: pipert2/core/base/wrappers/rpc_pipe_wrapper.py (hunk 13 above)
Comment (yarinshaked) [reply pipert2/core/base/wrappers/rpc_pipe_wrapper.py]: basof lo

**Review by tomerdayan168 — APPROVED** [review]
